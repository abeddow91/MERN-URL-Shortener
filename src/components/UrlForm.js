import React, {Component} from 'react';
import {isURL} from 'validator';
import axios from 'axios';

class urlForm extends Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleUrlSubmission = this.handleUrlSubmission.bind(this);

    this.state = {
      longUrl: '',
      shortUrl: '',
      shortId: 0,
      errorMessage: '',
    };
  }

  handleSubmit(e) {
    e.preventDefault();
    if (!isURL(this.state.longUrl)) {
      this.setState({
        longUrl: '',
        shortUrl: '',
        shortId: 0,
        errorMessage: 'Sorry that isn\'t a valid URL. Please provide a correct format eg: http://example.com'
      });
      this.refs.inputUrl.value= '' ;
      return;
    }
    this.handleUrlSubmission();
    this.setState({
      longUrl: '',
      errorMessage: '',
      shortUrl: 'localhost:8080/api/' + this.state.shortId
    });
    this.refs.inputUrl.value= '' ;
  }

  handleUrlSubmission(){
    var self = this;
    axios.post('/api/insert',
    {longUrl: self.state.longUrl})
    .then(response => {
      self.setState({
        shortId: response.data.short
      }, self.makeShortUrl);
    })
    .catch(function (err) {
      console.log('error', err);
    });
  }

  makeShortUrl() {
    var createdString = 'localhost:8080/api/' + this.state.shortId;
    this.setState({
      shortUrl: createdString
    });
  }

  handleChange(e) {
    this.setState({longUrl: e.target.value},
    );
  }

  render() {
    return (

      <div className="urls text-center">
      <form className='form' onSubmit={this.handleSubmit}>
      <br />
      <input className='input' ref='inputUrl' placeholder='Paste a URL to shorten' onChange={this.handleChange}/>
      <br />
      <br />
      <button className="btn btn-primary btn-lg">Shorten URL</button>
      <br />
      <br />
      {this.state.shortUrl.length > 0 ? 'Your new URL is: ' + this.state.shortUrl : this.state.errorMessage}
      </form>
      </div>
    );
  }
}

export default urlForm;
