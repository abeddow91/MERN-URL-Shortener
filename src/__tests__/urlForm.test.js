import React from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils from 'react-addons-test-utils';
import {shallow, mount} from 'enzyme';
import toJson from 'enzyme-to-json';

import UrlForm from '../components/UrlForm';

//This test mounts a component and checks that it doesn't throw an exception during rendering.

it('renders without crashing', () => {
  mount(<UrlForm />);
});

it('should render the Header', () => {
  const urlForm = shallow(<UrlForm />);
  const tree = toJson(urlForm);
  expect(tree).toMatchSnapshot()
});

it("it renders with expected state", () => {
  const testDoc = ReactTestUtils.renderIntoDocument(<UrlForm />);
  expect(testDoc.state.longUrl).toBe("");
  expect(testDoc.state.shortUrl).toBe("");
  expect(testDoc.state.shortId).toBe(0);
});

it("it stores the users input", () => {
  const testDoc = ReactTestUtils.renderIntoDocument(<UrlForm />);
  const input = ReactTestUtils.findRenderedDOMComponentWithClass(testDoc, "input");
  const btn = ReactTestUtils.findRenderedDOMComponentWithClass(testDoc, "btn");
  expect(testDoc.state.longUrl).toBe("");
  input.value = 'https://www.facebook.com';
  ReactTestUtils.Simulate.change(input);
  ReactTestUtils.Simulate.click(btn);
  expect(testDoc.state.longUrl).toBe('https://www.facebook.com');
});

it("it stores the error message if an incorrect url is submitted", () => {
  const testDoc = ReactTestUtils.renderIntoDocument(<UrlForm />);
  const input = ReactTestUtils.findRenderedDOMComponentWithClass(testDoc, "input");
  const btn = ReactTestUtils.findRenderedDOMComponentWithClass(testDoc, "btn");
  const form = ReactTestUtils.findRenderedDOMComponentWithClass(testDoc, "form");
  input.value = 'notaurl';
  ReactTestUtils.Simulate.change(input);
  ReactTestUtils.Simulate.submit(form);
  expect(testDoc.state.errorMessage).toBe('Sorry that isn\'t a valid URL. Please provide a correct format eg: http://example.com');
});

it("it clears the errorMessage if an correct url is submitted", () => {
  const testDoc = ReactTestUtils.renderIntoDocument(<UrlForm />);
  const input = ReactTestUtils.findRenderedDOMComponentWithClass(testDoc, "input");
  const btn = ReactTestUtils.findRenderedDOMComponentWithClass(testDoc, "btn");
  const form = ReactTestUtils.findRenderedDOMComponentWithClass(testDoc, "form");
  input.value = 'facebook';
  ReactTestUtils.Simulate.change(input);
  ReactTestUtils.Simulate.submit(form);    input.value = 'https://www.facebook.com';
  ReactTestUtils.Simulate.change(input);
  ReactTestUtils.Simulate.submit(form);
  expect(testDoc.state.errorMessage).toBe('');
});

it("it clears the longUrl after submitted", () => {
    const testDoc = ReactTestUtils.renderIntoDocument(<UrlForm />);
    const input = ReactTestUtils.findRenderedDOMComponentWithClass(testDoc, "input");
    const btn = ReactTestUtils.findRenderedDOMComponentWithClass(testDoc, "btn");
    const form = ReactTestUtils.findRenderedDOMComponentWithClass(testDoc, "form");
    input.value = 'https://www.facebook.com';
    ReactTestUtils.Simulate.change(input);
    ReactTestUtils.Simulate.submit(form);
    expect(testDoc.state.longUrl).toBe("");
  });

  it("creates a shortened URL and stores in state", () => {
    const testDoc = ReactTestUtils.renderIntoDocument(<UrlForm />);
    testDoc.state.shortId = 160291;
    testDoc.makeShortUrl();
    expect(testDoc.state.shortUrl).toBe('localhost:8080/api/160291');
  });
