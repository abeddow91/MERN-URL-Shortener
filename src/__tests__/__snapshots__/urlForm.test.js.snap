// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the Header 1`] = `
<div
  className="urls text-center"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      className="input"
      onChange={[Function]}
      placeholder="Paste a URL to shorten"
    />
    <br />
    <button
      className="btn btn-primary btn-lg"
    >
      Shorten URL
    </button>
  </form>
</div>
`;
